Add support to serialize/deserialize digest state for export/import

simo5 opened the pull request:
This PR adds support to export (and then import back) the state of a digest via custom serialization.

This addresses #14222 in the most simplest way I could think of.

It intentionally does not create a format that is portable across version, although it does not explicitly break the ability to import across version if the context state is not changed.

The code allow provider to do whatever they want when state is exported and it intentionally returns an opaque buffer of arbitrary length.

This is compatible with what providers like "pkcs11" can offer via the  C_GetOperationState()/C_SetOperationState() PKCS#11 calls.

There is a minimum of safeguard to avoid mixing the state of an implementation vs others by either using the hash length (for SHA2 functions) or a specially crafted "id" for KECCAK functions.

A magic string is provided so that attempt to restore a state saved with an incompatible version can be immediately recognized.

A saved state can be restore onto a valid EVP_MD_CTX of the same type at any time, and it completely replaces and obliterates any existing data on the context.

The "storage" format is intentionally kept as simple as possible to avoid the need to drag in things like ASN.1 parsing into it.

It is assumed that only trusted input is feed to the Deserialization interfaces, so no adversarial input sanitation is implemented.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

paulidale commented on providers/implementations/digests/sha2_prov.c:
```diff
@@ -56,41 +57,350 @@ static int sha1_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+                                                                              \
+        if (vctx == NULL)                                                     \
+            return 0;                                                         \
+        if ((p = OSSL_PARAM_locate(params, OSSL_DIGEST_SERIALIZATION)) != NULL) { \
```
Having just got rid of all the OSSL_PARAM_locate calls, I'd much prefer for new ones to not be introduced.  I.e. this should be converted to use a generated decoder.

paulidale commented on providers/implementations/digests/sha2_prov.c:
```diff
@@ -56,41 +57,350 @@ static int sha1_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+static OSSL_FUNC_digest_gettable_ctx_params_fn sha2_gettable_ctx_params;
+
+static const OSSL_PARAM known_sha2_settable_ctx_params[] = {
+    {OSSL_DIGEST_SERIALIZATION, OSSL_PARAM_OCTET_STRING, NULL, 0, 0},
```
Should serialisation/de-serialisatiob really be via a param?
I'd have thought they deserve provider API support and new EVP APIs to access them.

simo5 commented:
@paulidale how do we move this one forward?

paulidale commented:
I object to using parameters, so please don't merge.

Still waiting on a formal process.  There are some suggestions as to what to do.

slontis commented:
Do other @openssl/committers have opinions on this?

simo5 commented:
Ok I rebased the PR to add EVP_DigestSerialize/EVP_DigestDeserialize to handle the operation, and removed all of the params stuff.

I retained the conversion of the sha2_prov.c file to use include files for params, given the work was done

slontis commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -318,6 +321,27 @@ Similar to EVP_DigestFinalXOF() but allows multiple calls to be made to
+EVP_DigestDeserialize() deserializes the data from I<in> of size I<inlen>
+into the digest context I<ctx>.
+
+These functions does not guarantee importability of state exported by a
```
Need feedback from other @openssl/committers.
Is this ok? We dont allow incompatability across OpenSSL providers in most other places. Supporting a version number is not hard to do (I understand you are using it ephemerally, but is it possible that someone might want a persistent state?)

slontis commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -318,6 +321,27 @@ Similar to EVP_DigestFinalXOF() but allows multiple calls to be made to
+different providers.
+
+If a provider does not support serialization this function returns 0 and
+an an error code is set on the error stack. The L<ERR_get_error(3)> function
```
Is the error code bit necessary?

beldmit replied:
I don't think so

slontis replied:
The docs would be pretty bloated if we start doing this everywhere.

slontis reviewed (commented):
provider-digest.pod should be updated (update the history section also).

simo5 commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -318,6 +321,27 @@ Similar to EVP_DigestFinalXOF() but allows multiple calls to be made to
+EVP_DigestSerialize() serializes the state of the digest context I<ctx>
+and stores it in I<out>, unless I<out> is NULL.
+Otherwise the maximum necessary size of the output buffer is written to
+the I<outlen> parameter. If I<sig> is not NULL then before the call the
```
ouch

simo5 commented:
> Should we have a check that deserialization should work only for a clean target context?

I am not sure what a "clean target context" is, currently a deserialization always works as long as the context is initialized and not finalized, and overwrites any data in the context.

I checked to see if we have a way to indicate that the context has been initialized from the EVP API side, but the DigestInit context does not set any flag after calling the provider's initialization function.
That said the provider will fail to import if a minimum of initialization has not been done to set things like md_len in the various internal hash state structures.

I do not think we want to require that a context is squeaky clean though, it just needs to be consistent.

npajkovsky commented on doc/man3/EVP_DigestInit.pod:
```diff
@@ -65,6 +66,8 @@ EVP_MD_CTX_type, EVP_MD_CTX_pkey_ctx, EVP_MD_CTX_md_data
  int EVP_DigestFinalXOF(EVP_MD_CTX *ctx, unsigned char *out, size_t outlen);
  int EVP_DigestSqueeze(EVP_MD_CTX *ctx, unsigned char *out, size_t outlen);
+ int EVP_DigestSerialize(EVP_MD_CTX *ctx, unsigned char *out, size_t *outlen);
+ int EVP_DigestDeserialize(EVP_MD_CTX *ctx, const unsigned char *in, size_t inlen);
```
I'm for EVP_MD_CTX_serialize() and EVP_MD_CTX_deserialize() instead of CamelCase.

slontis commented on providers/implementations/digests/sha3_prov.c:
```diff
@@ -584,6 +587,127 @@ static void *keccak_dupctx(void *ctx)
+
+    /* buf */
+    memcpy(p, c->buf, sizeof(c->buf));
+    p += sizeof(c->buf);
```
remove the unused increment at the end of  all serialize functions

simo5 commented:
> What is test/pubkey.der?

garbage, removing

> paramnames.pm also seems strange?

I think my editor at some point decided to make the file executable bcause it is a script.

Ignore this noise, I am removing all of that

slontis commented on providers/implementations/digests/sha2_prov.c:
```diff
@@ -13,46 +13,253 @@
+
+    /* md_len check */
+    p = OPENSSL_load_u32_le(&val, p);
+    if ((unsigned int)val != c->md_len) {
```
NIT on coding style... single liners should not be in {}

slontis replied:
Looks like it duped in other functions

slontis commented on test/evp_extra_test2.c:
```diff
@@ -1394,6 +1394,75 @@ static int test_evp_md_ctx_copy2(void)
+        || !TEST_mem_eq(d1, d1_len, d2, d2_len))
+        goto end;
+
+    /* check that serialization fials on finalized contexts */
```
fails

slontis commented:
ping @openssl/committers

beldmit reviewed (changes requested):
I think that the format of the blob should also be documented.

Also there is a conflict in ‎CHANGES.md‎

beldmit reviewed (approved):
LGTM

simo5 commented on providers/implementations/digests/sha3_prov.c:
```diff
@@ -576,6 +579,126 @@ static void *keccak_dupctx(void *ctx)
 }
 
+static const unsigned char keccakmagic[] = "KECCAKv1";
+#define KACCAKMAGIC_LEN (sizeof(keccakmagic) - 1)
```
ugh

t8m commented on providers/implementations/digests/sha2_prov.c:
```diff
@@ -63,34 +269,48 @@ IMPLEMENT_digest_functions_with_settable_ctx(
+    SHA256_CBLOCK, SHA224_DIGEST_LENGTH,
+    SHA2_FLAGS, SHA224_Init,
+    SHA224_Update, SHA224_Final,
+    SHA256_Serialize, SHA256_Deserialize)
```
Ah, right.

beldmit reviewed (approved):
LGTM

openssl-machine commented:
This pull request is ready to merge
